src/webgpu/util/texture/texelData.spec.ts

```diff
@@ -0,0 +1,377 @@
+    components[c] === undefined && delete components[c];
+  });
+  doTest(t, t.params.format, components);
+};
+
+const makeParam = (
+  format: EncodableTextureFormat,
+  fn: (bitLength: number, index: number) => number
+) => {
+  const rep = getTexelDataRepresentation(format);
+  return {
+    R: rep.componentInfo.R ? fn(rep.componentInfo.R.bitLength, 0) : undefined,
+    G: rep.componentInfo.G ? fn(rep.componentInfo.G.bitLength, 1) : undefined,
+    B: rep.componentInfo.B ? fn(rep.componentInfo.B.bitLength, 2) : undefined,
+    A: rep.componentInfo.A ? fn(rep.componentInfo.A.bitLength, 3) : undefined,
+  };
+};
+
+g.test('unorm_texel_data_in_shader')
+  .params(
+    params()
+      .combine(poptions('format', kEncodableTextureFormats))
+      .filter(({ format }) => {
+        return (
+          kEncodableTextureFormatInfo[format].copySrc &&
+          kEncodableTextureFormatInfo[format].copyDst &&
+          kEncodableTextureFormatInfo[format].color &&
+          kEncodableTextureFormatInfo[format].dataType === 'unorm'
+        );
+      })
+      .expand(({ format }) => {
+        const max = (bitLength: number) => Math.pow(2, bitLength) - 1;
+        return [
+          // Test extrema
+          makeParam(format, () => 0),
+          makeParam(format, bitLength => max(bitLength)),
+
+          // Test a middle value
+          makeParam(format, bitLength => Math.floor(max(bitLength) / 2)),
+
+          // Test mixed values
+          makeParam(format, (bitLength, i) => {
+            const offset = [0.13, 0.63, 0.42, 0.89];
+            return Math.floor(offset[i] * max(bitLength));
+          }),
+        ];
+      })
+  )
+  .fn(testFn);
+
+g.test('snorm_texel_data_in_shader')
+  .params(
+    params()
+      .combine(poptions('format', kEncodableTextureFormats))
+      .filter(({ format }) => {
+        return (
+          kEncodableTextureFormatInfo[format].copySrc &&
+          kEncodableTextureFormatInfo[format].copyDst &&
+          kEncodableTextureFormatInfo[format].color &&
+          kEncodableTextureFormatInfo[format].dataType === 'snorm'
+        );
+      })
+      .expand(({ format }) => {
+        const max = (bitLength: number) => Math.pow(2, bitLength - 1) - 1;
+        return [
+          // Test extrema
+          makeParam(format, () => 0),
+          makeParam(format, bitLength => max(bitLength)),
+          makeParam(format, bitLength => -max(bitLength)),
+          makeParam(format, bitLength => -max(bitLength) - 1),
+
+          // Test a middle value
+          makeParam(format, bitLength => Math.floor(max(bitLength) / 2)),
+
+          // Test mixed values
+          makeParam(format, (bitLength, i) => {
+            const offset = [0.13, 0.63, 0.42, 0.89];
+            const range = 2 * max(bitLength);
+            return -max(bitLength) + Math.floor(offset[i] * range);
+          }),
+        ];
+      })
+  )
+  .fn(testFn);
+
+g.test('uint_texel_data_in_shader')
```

Comment:
Could be, but I think `sint` and `snorm` should be separate at least since snorm tests -128 and -127 whereas sint cares only about -128. And if they're separate I'd rather keep uint and unorm separate for consistency.